packages/account/src/BiconomySmartAccountV2.ts

```diff
@@ -499,11 +500,66 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
+   * @param {PaymasterUserOperationDto} paymasterServiceData - The paymaster service data containing mode and additional information.
+   * @returns {Promise<Partial<UserOperationStruct>>} A promise that resolves to the modified user operation structure.
+   */
+  async setPaymasterUserOp(
+    userOp: Partial<UserOperationStruct>,
+    paymasterServiceData: PaymasterUserOperationDto,
+  ): Promise<Partial<UserOperationStruct>> {
+    if (this.paymaster !== undefined) {
+      try {
+        if (paymasterServiceData.mode === PaymasterMode.SPONSORED) {
+          const paymasterData = await (this.paymaster as IHybridPaymaster<PaymasterUserOperationDto>).getPaymasterAndData(userOp, {
+            mode: paymasterServiceData.mode,
+          });
+          userOp.paymasterAndData = paymasterData.paymasterAndData;
+          userOp.callGasLimit = paymasterData.callGasLimit;
+          userOp.verificationGasLimit = paymasterData.verificationGasLimit;
+          userOp.preVerificationGas = paymasterData.preVerificationGas;
+          return userOp;
+        } else if (paymasterServiceData.mode === PaymasterMode.ERC20 && paymasterServiceData.feeQuote !== undefined) {
+          const finalUserOp = await this.buildTokenPaymasterUserOp(userOp, {
+            feeQuote: paymasterServiceData.feeQuote,
+            spender: (paymasterServiceData.spender as Hex) || "",
+            maxApproval: paymasterServiceData.maxApproval,
+          });
+          const newPaymasterServiceData = {
+            mode: PaymasterMode.ERC20,
+            feeTokenAddress: paymasterServiceData.feeQuote.tokenAddress,
+            calculateGasLimits: true, // Always recommended and especially when using token paymaster
+          };
+          const paymasterAndDataWithLimits = await (this.paymaster as IHybridPaymaster<PaymasterUserOperationDto>).getPaymasterAndData(
+            finalUserOp,
+            newPaymasterServiceData,
+          );
+          finalUserOp.paymasterAndData = paymasterAndDataWithLimits.paymasterAndData;
+          finalUserOp.callGasLimit = paymasterAndDataWithLimits.callGasLimit;
+          finalUserOp.verificationGasLimit = paymasterAndDataWithLimits.verificationGasLimit;
+          finalUserOp.preVerificationGas = paymasterAndDataWithLimits.preVerificationGas;
+          return finalUserOp;
+        } else {
+          return userOp;
+        }
+      } catch (e: any) {
+        Logger.error("Error while fetching paymaster data", e);
+        return userOp;
+      }
+    } else {
+      Logger.error("Paymaster is not provided");
+      return userOp;
+    }
+  }
+
   /**
    *
    * @param userOp
    * @param params
-   * @description This function call will take 'unsignedUserOp' as an input, sign it with the owner key, and send it to the bundler.
+   * @description This function call will setup paymaster values (if paymaster is used), take 'unsignedUserOp' as an input, sign it with the owner key, and send it to the bundler.
```

Comment:
the implementation here doesn't look like it will setup paymaster values

Reply:
Fixed